docs/writing-controllers.md

```diff
@@ -0,0 +1,1447 @@
+- They act as a communication layer between service layers (blockchains, internal or external APIs, etc.).
+- They allow the application to be divided into logical modules which can be maintained by different teams within the company.
+
+## Use the latest version of `BaseController` for controllers
+
+All controllers should inherit from `BaseController` from the `@metamask/base-controller` package. This provides a few benefits:
+
+- It defines a standard interface for all controllers.
+- It introduces the messenger system, which is useful for interacting with other controllers without requiring direct access to them.
+- It enforces that `update` is the only way to modify the state of the controller and provides a way to listen for state updates via the messenger.
+- It simplifies initialization by consolidating constructor arguments into one.
+
+## Don't use `BaseController` for non-controllers
+
+One of the uniquely identifying features of a controller is the ability to manage state.
+
+If you have a class that does not capture any data in state, then your class does not need to inherit from `BaseController` (even if it uses the messaging system).
+
+## Maintain a clear and concise API
+
+The name of the controller should reflect its responsibility. If, when creating a controller, it is difficult to come up with a good name, consider defining the responsibility first.
+
+Each public method and each state property of a controller should have a purpose, and the name of the method or state property should be readable and should reflect the purpose clearly. If something does not need to be public, it should be made private; if it is unnecessary, it should be removed.
+
+## Accept a optional, partial representation of state
+
+Although `BaseController` requires a full representation of controller state, in practice, controllers should accept a partial version and then supply missing properties with defaults. In fact, the `state` argument should be optional:
+
+```typescript
+type FooControllerState = {
+  // ...
+};
+
+function getDefaultFooControllerState(): FooControllerState {
+  // ...
+}
+
+class FooController extends BaseController</* ... */> {
+  constructor({
+    messenger,
+    state = {},
+  }: {
+    messenger: FooControllerMessenger;
+    state: Partial<FooControllerState>;
+  }) {
+    super({
+      // ...
+      messenger,
+      state: { ...getDefaultFooControllerState(), ...state },
+    });
+
+    // ...
+  }
+}
+```
+
+## Provide a default representation of state
+
+Each controller needs a default representation in order to fully initialize itself when [receiving a partial representation of state](#accept-a-partial-representation-of-state). A default representation of state is also useful when testing interactions with a controller's `*:stateChange` event.
+
+It is customary to export a function which returns this default representation:
```

Comment:
nit: Providing an explanation for why we prefer a function might make it easier to encourage this practice.

```suggestion
A function which returns this default representation should be defined and exported. 

Note that the default representation object is not exported directly to ensure that we are always providing a newly initialized instance with a new object reference.
```

Reply:
Good idea!